File: api/app/parsers/package_lock.py
import json
from pathlib import Path
from typing import List

from app.parsers.base import register_parser


@register_parser("package-lock.json", "npm")
def parse_package_lock(file_path: Path) -> List[tuple[str, str, str]]:
    dependencies = []
    with file_path.open() as file:
        data = json.load(file)
        for name, info in data.get("dependencies", {}).items():
            version = info.get("version", "unknown")
            dependencies.append((name, version, "npm"))

    return dependencies

File: background-worker/internal/parsers/package_lock.go
package parsers

import (
	"encoding/json"
	"fmt"
	"os"
)

func init() {
	RegisterParser("package-lock.json", "npm", ParsePackageLock)
}

func ParsePackageLock(path string) ([]Dependency, error) {
	file, err := os.Open(path)

	if err != nil {
		return nil, err
	}

	defer func() {
		if err := file.Close(); err != nil {
			fmt.Fprintf(os.Stderr, "failed to close file: %v\n", err)
		}
	}()

	var raw map[string]any

	if err := json.NewDecoder(file).Decode(&raw); err != nil {
		return nil, err
	}

	deps := []Dependency{}

	if dependencies, ok := raw["dependencies"].(map[string]any); ok {
		for name, val := range dependencies {
			if info, ok := val.(map[string]any); ok {
				version := "unknown"

				if v, ok := info["version"].(string); ok {
					version = v
				}

				deps = append(deps, Dependency{
					Name:      name,
					Version:   version,
					Ecosystem: "npm",
				})
			}
		}
	}

	return deps, nil
}
